lib/icinga/dependency-group.cpp

```diff
@@ -0,0 +1,362 @@
+		AddMember(dependency);
+	}
+}
+
+/**
+ * Create a composite key for the provided dependency.
+ *
+ * @param dependency The dependency object to create a composite key for.
+ *
+ * @return - Returns the composite key for the provided dependency.
+ */
+DependencyGroup::CompositeKeyType DependencyGroup::MakeCompositeKeyFor(const Dependency::Ptr& dependency)
+{
+	return std::make_tuple(
+		dependency->GetParent()->GetName(),
+		dependency->GetPeriodRaw(),
+		dependency->GetStateFilter(),
+		dependency->GetIgnoreSoftStates()
+	);
+}
+
+/**
+ * Check if the current dependency group has any members.
+ *
+ * @return bool - Returns true if the current dependency group has any members.
+ */
+bool DependencyGroup::HasMembers() const
+{
+	std::lock_guard lock(m_Mutex);
+	return !m_Members.empty();
+}
+
+/**
+ * Check whether an identical dependency object is member of the current dependency group.
+ *
+ * Identical means any dependency object that produces the exact same key (CompositeKeyType) as the provided one.
+ *
+ * @param dependency The dependency to look for.
+ *
+ * @return bool - Returns true if the provided dependency is member of the current dependency group.
+ */
+bool DependencyGroup::HasIdenticalMember(const Dependency::Ptr& dependency) const
+{
+	std::lock_guard lock(m_Mutex);
+	return m_Members.find(MakeCompositeKeyFor(dependency)) != m_Members.end();
+}
+
+/**
+ * Retrieve all members of the current dependency group the provided child Checkable depend on.
+ *
+ * Note, in order to ease duplicated dependencies exhaustion, the returned members are sorted by the parent Checkable.
+ * This way, all identical dependencies are placed next to each other, and you can easily consume them via a simple loop.
+ *
+ * @param child The child Checkable to look for.
+ *
+ * @return - Returns all members of the current dependency group the provided child depend on.
+ */
+std::vector<Dependency::Ptr> DependencyGroup::GetMembers(const Checkable* child) const
+{
+	std::lock_guard lock(m_Mutex);
+	std::vector<Dependency::Ptr> members;
+	for (auto& [_, dependencies] : m_Members) {
+		auto [begin, end] = dependencies.equal_range(child);
+		std::transform(begin, end, std::back_inserter(members), [](const auto& member) {
+			return member.second;
+		});
+	}
+
+	std::sort(members.begin(), members.end(), [](const Dependency::Ptr& lhs, const Dependency::Ptr& rhs) {
+		return lhs->GetParent() < rhs->GetParent();
+	});
+	return members;
+}
+
+/**
+ * Load all parent Checkables of the current dependency group members.
+ *
+ * @param parents The set to load the parent Checkables into.
+ */
+void DependencyGroup::LoadParents(std::set<Checkable::Ptr>& parents) const
+{
+    std::lock_guard lock(m_Mutex);
+    for (auto& [_, dependencies] : m_Members) {
+    	VERIFY(!dependencies.empty()); // We should never have an empty map for any given key at any given time.
+    	parents.insert(dependencies.begin()->second->GetParent());
+    }
```

Comment:
If the key would use pointers instead of strings (see previous comment), that could then simply fetch the parents from the key and avoid the `VERIFY()` (could still be kept as a debug-only `ASSERT()` though).